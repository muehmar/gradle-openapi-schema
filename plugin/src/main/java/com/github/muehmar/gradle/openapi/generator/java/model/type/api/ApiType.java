package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.ParameterizedApiClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeConversion;
import lombok.Value;

@Value
public class ApiType {
  QualifiedClassName className;
  ParameterizedApiClassName parameterizedClassName;
  ToApiTypeConversion toApiTypeConversion;
  FromApiTypeConversion fromApiTypeConversion;

  public static ApiType fromConversion(
      QualifiedClassName className, TypeConversion conversion, PList<JavaType> generics) {
    final ParameterizedApiClassName parameterizedClassName =
        ParameterizedApiClassName.ofClassNameAndGenerics(className, generics);
    final ToApiTypeConversion toApiTypeConversion =
        new ToApiTypeConversion(ConversionMethod.ofString(className, conversion.getToCustomType()));
    final FromApiTypeConversion fromApiTypeConversion =
        new FromApiTypeConversion(
            ConversionMethod.ofString(className, conversion.getFromCustomType()));
    return new ApiType(
        className, parameterizedClassName, toApiTypeConversion, fromApiTypeConversion);
  }
}
