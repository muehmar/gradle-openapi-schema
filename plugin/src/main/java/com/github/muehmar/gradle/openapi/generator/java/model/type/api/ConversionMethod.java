package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class ConversionMethod {
  private final FactoryMethodConversion factoryMethodConversion;
  private final InstanceMethodConversion instanceMethodConversion;

  private ConversionMethod(
      FactoryMethodConversion factoryMethodConversion,
      InstanceMethodConversion instanceMethodConversion) {
    this.factoryMethodConversion = factoryMethodConversion;
    this.instanceMethodConversion = instanceMethodConversion;
  }

  public static ConversionMethod ofString(String conversionMethod) {
    return FactoryMethodConversion.fromString(conversionMethod)
        .map(ConversionMethod::ofFactoryMethod)
        .orElseGet(
            () ->
                ConversionMethod.ofInstanceMethod(
                    InstanceMethodConversion.ofString(conversionMethod)));
  }

  static ConversionMethod ofFactoryMethod(FactoryMethodConversion factoryMethodConversion) {
    return new ConversionMethod(factoryMethodConversion, null);
  }

  static ConversionMethod ofInstanceMethod(InstanceMethodConversion instanceMethodConversion) {
    return new ConversionMethod(null, instanceMethodConversion);
  }

  public <T> T fold(
      Function<FactoryMethodConversion, T> onFactoryMethod,
      Function<InstanceMethodConversion, T> onInstanceMethod) {
    if (factoryMethodConversion != null) {
      return onFactoryMethod.apply(factoryMethodConversion);
    } else {
      return onInstanceMethod.apply(instanceMethodConversion);
    }
  }
}
