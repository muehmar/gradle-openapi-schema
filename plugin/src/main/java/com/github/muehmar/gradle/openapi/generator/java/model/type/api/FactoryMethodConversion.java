package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import java.util.Optional;
import lombok.Value;

@Value
public class FactoryMethodConversion {
  QualifiedClassName className;
  Name methodName;

  public static Optional<FactoryMethodConversion> fromString(
      QualifiedClassName className, String factoryMethodConversion) {
    final String[] parts = factoryMethodConversion.split("#");
    if (parts.length != 2 && factoryMethodConversion.contains("#")) {
      throw createInvalidFormatException(factoryMethodConversion);
    }
    if (parts.length < 2) {
      return Optional.empty();
    }

    try {
      final QualifiedClassName parsedClassName = QualifiedClassName.ofQualifiedClassName(parts[0]);
      final Name methodName = Name.ofString(parts[1]);
      if (not(parsedClassName.getPackageName().isPresent())
          && parsedClassName.getClassName().equals(className.getClassName())) {
        return Optional.of(new FactoryMethodConversion(className, methodName));
      }
      return Optional.of(new FactoryMethodConversion(parsedClassName, methodName));
    } catch (Exception e) {
      throw createInvalidFormatException(factoryMethodConversion);
    }
  }

  private static OpenApiGeneratorException createInvalidFormatException(
      String factoryMethodConversion) {
    return new OpenApiGeneratorException(
        "Invalid configuration for type conversion '%s'. Expected format for static factory methods: 'package.name.ClassName#methodName'.",
        factoryMethodConversion);
  }
}
