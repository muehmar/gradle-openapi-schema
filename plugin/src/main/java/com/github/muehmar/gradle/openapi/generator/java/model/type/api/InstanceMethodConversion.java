package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import lombok.Value;

@Value
public class InstanceMethodConversion {
  Name methodName;

  public static InstanceMethodConversion ofString(String methodName) {
    return new InstanceMethodConversion(Name.ofString(methodName));
  }
}
