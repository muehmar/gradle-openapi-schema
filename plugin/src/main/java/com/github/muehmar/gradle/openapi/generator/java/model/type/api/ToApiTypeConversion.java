package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import java.util.function.Function;
import lombok.Value;

@Value
public class ToApiTypeConversion {
  ConversionMethod conversionMethod;

  public ToApiTypeConversion(ConversionMethod conversionMethod) {
    this.conversionMethod = conversionMethod;
  }

  public <T> T fold(
      Function<FactoryMethodConversion, T> onFactoryMethod,
      Function<InstanceMethodConversion, T> onInstanceMethod) {
    return conversionMethod.fold(onFactoryMethod, onInstanceMethod);
  }
}
