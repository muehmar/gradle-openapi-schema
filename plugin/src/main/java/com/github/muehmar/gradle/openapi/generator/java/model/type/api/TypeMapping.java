package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMapping;
import com.github.muehmar.gradle.openapi.generator.settings.FormatTypeMapping;
import java.util.Optional;
import lombok.Value;

@Value
public class TypeMapping {
  QualifiedClassName className;
  Optional<ApiType> apiType;

  public static TypeMapping fromClassName(QualifiedClassName className) {
    return new TypeMapping(className, Optional.empty());
  }

  public static TypeMapping fromClassMappings(
      QualifiedClassName originalClassName, PList<ClassTypeMapping> classTypeMappings) {
    return fromClassMappings(originalClassName, classTypeMappings, PList.empty());
  }

  public static TypeMapping fromClassMappings(
      QualifiedClassName originalClassName,
      PList<ClassTypeMapping> classTypeMappings,
      PList<JavaType> generics) {
    return classTypeMappings
        .filter(
            classMapping ->
                classMapping.getFromClass().equals(originalClassName.getClassName().asString()))
        .headOption()
        .map(classMapping -> fromClassMapping(originalClassName, classMapping, generics))
        .orElseGet(() -> fromClassName(originalClassName));
  }

  private static TypeMapping fromClassMapping(
      QualifiedClassName originalClassName,
      ClassTypeMapping classTypeMapping,
      PList<JavaType> generics) {
    final QualifiedClassName mappedClassName =
        QualifiedClassName.fromClassTypeMapping(classTypeMapping);

    return classTypeMapping
        .getTypeConversion()
        .map(conversion -> ApiType.fromConversion(mappedClassName, conversion, generics))
        .map(apiType -> new TypeMapping(originalClassName, Optional.of(apiType)))
        .orElseGet(() -> fromClassName(mappedClassName));
  }

  public static TypeMapping fromFormatMappings(
      QualifiedClassName originalClassName,
      String formatString,
      PList<FormatTypeMapping> formatTypeMappings) {
    return fromFormatMappings(originalClassName, formatString, formatTypeMappings, PList.empty());
  }

  public static TypeMapping fromFormatMappings(
      QualifiedClassName originalClassName,
      String formatString,
      PList<FormatTypeMapping> formatTypeMappings,
      PList<JavaType> generics) {
    return formatTypeMappings
        .filter(formatMapping -> formatMapping.getFormatType().equals(formatString))
        .headOption()
        .map(formatMapping -> fromFormatMapping(originalClassName, formatMapping, generics))
        .orElseGet(() -> fromClassName(originalClassName));
  }

  private static TypeMapping fromFormatMapping(
      QualifiedClassName originalClassName,
      FormatTypeMapping formatTypeMapping,
      PList<JavaType> generics) {
    final QualifiedClassName mappedClassName =
        QualifiedClassName.ofQualifiedClassName(formatTypeMapping.getClassType());

    return formatTypeMapping
        .getTypeConversion()
        .map(conversion -> ApiType.fromConversion(mappedClassName, conversion, generics))
        .map(apiType -> new TypeMapping(originalClassName, Optional.of(apiType)))
        .orElseGet(() -> fromClassName(mappedClassName));
  }

  public TypeMapping or(TypeMapping other, QualifiedClassName originalClassName) {
    if (this.apiType.isPresent()) {
      return this;
    } else if (other.getApiType().isPresent()) {
      return other;
    } else if (this.className.equals(originalClassName)) {
      return other;
    } else {
      return this;
    }
  }
}
