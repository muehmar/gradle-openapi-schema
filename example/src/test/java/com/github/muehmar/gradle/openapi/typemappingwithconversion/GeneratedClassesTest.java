package com.github.muehmar.gradle.openapi.typemappingwithconversion;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import com.github.muehmar.gradle.openapi.util.ClassFinder;
import java.util.List;
import java.util.stream.Collectors;
import org.junit.jupiter.api.Test;

@SnapshotTest
class GeneratedClassesTest {
  private Expect expect;

  @Test
  @SnapshotName("allGeneratedClasses")
  void findNonTestClassesInPackage_when_modelPackage_then_allClasses() {
    final List<String> classNames =
        ClassFinder.findNonTestClassesInPackage(getClass().getPackage().getName()).stream()
            .filter(className -> !className.equals("CustomStrings"))
            .filter(className -> !className.equals("CustomString"))
            .filter(className -> !className.equals("CustomList"))
            .collect(Collectors.toList());

    expect.toMatchSnapshot(String.join("\n", classNames));
  }
}
