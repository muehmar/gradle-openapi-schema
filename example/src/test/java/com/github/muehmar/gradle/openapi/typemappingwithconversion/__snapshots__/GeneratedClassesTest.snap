allGeneratedClasses=[
AllOfListObjectAllOfDto
AllOfListObjectDto
AllOfMapObjectAllOfDto
AllOfMapObjectDto
AnyOfListObjectAnyOfContainerDto
AnyOfListObjectAnyOfDto
AnyOfListObjectDto
AnyOfMapObjectAnyOfContainerDto
AnyOfMapObjectAnyOfDto
AnyOfMapObjectDto
ArrayAdditionalPropertiesDto
ArrayAdditionalPropertiesPropertyDto
ListObjectDto
MapObjectDto
OneOfListObjectDto
OneOfListObjectOneOfContainerDto
OneOfListObjectOneOfDto
OneOfMapObjectDto
OneOfMapObjectOneOfContainerDto
OneOfMapObjectOneOfDto
PosologyDto
UserDto
]